### PR TITLE
Move all enums to a single place

### PR DESCRIPTION
**Pull request by ashfall**
This is to avoid circular import issues described in #72. I tried to refactor without moving too many things around, and this sounded like the sanest thing to do. 

- I left `tls.ciphersuites` where they are, and am still debating if that's the right call. On the one hand, I feel like `CipherSuites` deserve their own module, which is why we didn't define them in `tls._constructs` initially. On the other hand, if the idea is to consolidate all enums in one place, I am breaking consistency by not putting `CipherSuites` in that one place. :|

- Something to consider: Should we have a `tls.helpers` that has, in it, what's now  `tls.enums` (as `tls.helpers._enums`) (and perhaps `tls.utils` -> `tls.helpers._utils`? Unsure if it actually fits in there), and an `__init__.py` that imports everything from `_enums`, so they can be used as `from tls.helpers import ThisCustomThing` throughout? Would that be a useful change, or should we maintain a distinction between what kind of helper we are defining?

- If anyone has a better name for `tls.enums`, please speak up! :)

### CODE REVIEW

**Comment by ashfall** [comment]
@glyph Rumor has it you may have opinions on this. Thoughts?

**Comment by markrwilliams** [comment]
No comments on the code because of how straightforward a refactor this is.  

> I left tls.ciphersuites where they are, and am still debating if that's the right call. On the one hand, I feel like CipherSuites deserve their own module, which is why we didn't define them in tls._constructs initially. On the other hand, if the idea is to consolidate all enums in one place, I am breaking consistency by not putting CipherSuites in that one place. :|

I think that's a fine call for this PR, the primary aim of which seems to be closing #72.  In fact, I'm betting the location and nature of `ciphersuite` will become clearer when we add a handshaking finite state machine.  So until at least then I think it's fine where it is and that this is ready to merge.

Everything that comes next is just for discussion, but only if what it adds to that discussion is helpful.  I don't intend to start bikeshedding!  Whatever you decide will be good.

> Something to consider: Should we have a tls.helpers that has, in it, what's now tls.enums (as tls.helpers._enums) (and perhaps tls.utils -> tls.helpers._utils? Unsure if it actually fits in there), and an __init__.py that imports everything from _enums, so they can be used as from tls.helpers import ThisCustomThing throughout? Would that be a useful change, or should we maintain a distinction between what kind of helper we are defining?

I like this idea!  Collecting these modules under a `helpers` package makes it clear their code supports the concrete implementation, and prepares the reader (or at least *this* reader) for private APIs and more abstract tests.  I think I'd like it to be `_helpers` so that it's clear that it's for internal use only.

I'm less excited by bubbling up imports from the sub-modules, but only because I prefer that approach for public APIs, while for internal/private APIs I prefer import paths that are as complete as possible.  I've found as a new reader it helps me find my way around a code base.  Yet moving imports into `_helpers.__init__` would free us up to shuffle things around inside `_helpers` without having to rewrite all the imports.  That seems OK, so I'll say that either way works for me.

However, I think I'd prefer it be named `_common` because the import would read nicely: `from .common import enums` or `from ._common._enums import HashAlgorithm`.  Also, while this code helps other code, it's actually code that's *common* to `tls` submodules. But that may be a bit much -- `_helpers` is certainly fine!

I'd also prefer that `utils` get a new name.  It doesn't contain utilities -- it contains abstract `construct`s.  By moving it under a `_helpers` or _common` package, it can be renamed `_constructs` or `constructs` without clashing with `tls._constructs.`   

> If anyone has a better name for tls.enums, please speak up! :)

I think `_helpers` will address this by enclosing `enums` in a more descriptive namespace.

Thanks for reading, and thanks a *ton* for the PR!!!  Issues keep getting closed.  Very exciting!

**Comment by sigmavirus24** [comment]
So I'll offer the other side of the discussion from Mark.

Flake8 recently refactored into sub-modules like that and while it works well for us, we have a lot of moving parts so having discrete sub-packages works for us. We don't put everything into the top level module `__init__` though because those packages are all meant as implementation details (with a couple exceptions) and not meant to be used by people using Flake8 or extending it. We have one place that we keep everything that should be importable (`flake8.api`) because Flake8 is first and foremost a command-line application.

That said, with my Requests hat on, if you want to mark something as internal only, documentation alone won't help. Even placing an `_` before the name doesn't help. Developers will use literally everything they can find. Using the `_` is conventional and will discourage some, so that might be good.

As for a `tls._helpers` it might be nice to have similar internal helper things in there. I'm not sure, though, that it is the best place to collect `enums`, `utils`, and other things. If these are truly internal details, you might want a more descriptive name but only to convey what exactly is in that package.

Also, for me, I would prefer you import the enums module (from wherever) and use the enums as `enums.<EnumName>.<EnumAttr>`. It's very clear that what you're using then is an Enum. If you shove `enums` into another module, I think this will be clearer as to what you're using actually is.

**Comment by ashfall** [comment]
In reply to @markrwilliams :
> No comments on the code because of how straightforward a refactor this is.  
> 
> > I left tls.ciphersuites where they are, and am still debating if that's the right call. On the one hand, I feel like CipherSuites deserve their own module, which is why we didn't define them in tls._constructs initially. On the other hand, if the idea is to consolidate all enums in one place, I am breaking consistency by not putting CipherSuites in that one place. :| 
> 
> I think that's a fine call for this PR, the primary aim of which seems to be closing #72.  In fact, I'm betting the location and nature of `ciphersuite` will become clearer when we add a handshaking finite state machine.  

Good point!

> So until at least then I think it's fine where it is and that this is ready to merge.

> 
> Everything that comes next is just for discussion, but only if what it adds to that discussion is helpful.  I don't intend to start bikeshedding!  Whatever you decide will be good.

Definitely helpful, thanks for taking the time to review and discuss!

> 
> > Something to consider: Should we have a tls.helpers that has, in it, what's now tls.enums (as tls.helpers._enums) (and perhaps tls.utils -> tls.helpers._utils? Unsure if it actually fits in there), and an __init__.py that imports everything from _enums, so they can be used as from tls.helpers import ThisCustomThing throughout? Would that be a useful change, or should we maintain a distinction between what kind of helper we are defining?
> 
> I like this idea!  Collecting these modules under a `helpers` package makes it clear their code supports the concrete implementation, and prepares the reader (or at least *this* reader) for private APIs and more abstract tests.  I think I'd like it to be `_helpers` so that it's clear that it's for internal use only.
> 
> I'm less excited by bubbling up imports from the sub-modules, but only because I prefer that approach for public APIs, while for internal/private APIs I prefer import paths that are as complete as possible.  I've found as a new reader it helps me find my way around a code base.  Yet moving imports into `_helpers.__init__` would free us up to shuffle things around inside `_helpers` without having to rewrite all the imports.  That seems OK, so I'll say that either way works for me.

Agreed. If I had to decide between the freedom to shuffle things around easily and helping new readers find their way around easily, I think I'll lean towards the latter.

> 
> However, I think I'd prefer it be named `_common` because the import would read nicely: `from .common import enums` or `from ._common._enums import HashAlgorithm`.  Also, while this code helps other code, it's actually code that's *common* to `tls` submodules. But that may be a bit much -- `_helpers` is certainly fine!

Makes sense, I like `common`. +1 to that.

> 
> I'd also prefer that `utils` get a new name.  It doesn't contain utilities -- it contains abstract `construct`s.  By moving it under a `_helpers` or _common` package, it can be renamed `_constructs` or `constructs` without clashing with `tls._constructs.`   

+1 on renaming `utils`. -1 on having more than one `_constructs` around. Thinking further, I think one of these (i.e. the current `tls._constructs` or the potential `tls.common.constructs` should be named something else. Thoughts? (Naming is hard :| ))

> 
> > If anyone has a better name for tls.enums, please speak up! :)
> 
> I think `_helpers` will address this by enclosing `enums` in a more descriptive namespace.

Agreed!
> 
> Thanks for reading, and thanks a *ton* for the PR!!!  Issues keep getting closed.  Very exciting!

Thanks a lot for your enthusiastic and valuable contributions!

**Comment by ashfall** [comment]
To @sigmavirus24:
 
> So I'll offer the other side of the discussion from Mark.

Ooh another reviewer! Thank you, and I'm happy to see you here :)

> 
> Flake8 recently refactored into sub-modules like that and while it works well for us, we have a lot of moving parts so having discrete sub-packages works for us. We don't put everything into the top level module `__init__` though because those packages are all meant as implementation details (with a couple exceptions) and not meant to be used by people using Flake8 or extending it. We have one place that we keep everything that should be importable (`flake8.api`) because Flake8 is first and foremost a command-line application.
> 
> That said, with my Requests hat on, if you want to mark something as internal only, documentation alone won't help. Even placing an `_` before the name doesn't help. Developers will use literally everything they can find. Using the `_` is conventional and will discourage some, so that might be good.

Hm, honestly, I am temped to not use `_` at all for the submodule name. I agree with what @markrwilliams said:
>  I think I'd like it to be _helpers so that it's clear that it's for internal use only.

but I'm having trouble thinking of a future use-case where using _everything_ inside `common` _will_ most definitely be dangerous, and I am not sure if making that assumption not will be OK in future, or if we'll regret it and have to refactor again :\  I don't have a good answer yet, I'm going to think on this a bit, appreciate more inputs / thoughts, or if anyone would like to convince me one way or the other, that would be awesome! :)

> 
> As for a `tls._helpers` it might be nice to have similar internal helper things in there. I'm not sure, though, that it is the best place to collect `enums`, `utils`, and other things. If these are truly internal details, you might want a more descriptive name but only to convey what exactly is in that package.
> 

True, I think `common` works better here, e.g. `tls.common.enums` are "common enums for tls".


> Also, for me, I would prefer you import the enums module (from wherever) and use the enums as `enums.<EnumName>.<EnumAttr>`. It's very clear that what you're using then is an Enum. If you shove `enums` into another module, I think this will be clearer as to what you're using actually is.

This is an excellent point, +1 to making the path clearer. I'm going to make that change and push a couple of commits.

Thanks again for your inputs, these are extremely helpful. (Please stick around! 😃  )

**Comment by sigmavirus24** [comment]
> I'm having trouble thinking of a future use-case where using everything inside common will most definitely be dangerous, and I am not sure if making that assumption will be OK in future, or if we'll regret it and have to refactor again :\

So, here's the thing though: Private implementation details don't need to be dangerous in order to require being hidden. Providing a small surface are API is good. *If* you find that things that are presently not exposed should be exposed, refactoring a little to do that isn't the worst thing ever. You're *intentionally* choosing at that point to make the API contract with your users. Going the opposite direction, however, is incredibly hard and something I'm certain @glyph would agree should be done as *rarely* as possible.

So if you tend towards keeping things "private" (as best you can achieve that in Python) then you'll have control over what your API contract is with your user.

You can always simply do this with documentation and not use naming practices. That's actually what Flake8 is choosing to do: https://gitlab.com/pycqa/flake8/tree/master/src/flake8.

**Comment by ashfall** [comment]
> So, here's the thing though: Private implementation details don't need to be dangerous in order to require being hidden. Providing a small surface are API is good. If you find that things that are presently not exposed should be exposed, refactoring a little to do that isn't the worst thing ever. You're intentionally choosing at that point to make the API contract with your users. 

That does make sense, thanks. I'm convinced to go with `_common`.

File: tls/test/test_message.py [hunk 1]
```diff
@@ -60,12 +61,14 @@ class TestCertificateRequestParsing(object):
...
     def test_parse_certificate_request(self):
         record = CertificateRequest.from_bytes(self.no_authorities_packet)
-        assert record.certificate_types == [ClientCertificateType.RSA_SIGN]
+        assert record.certificate_types == [
+            enums.ClientCertificateType.RSA_SIGN
+        ]
         assert len(record.supported_signature_algorithms) == 1
         assert record.supported_signature_algorithms[0].hash == \
-            HashAlgorithm.MD5
+            enums.HashAlgorithm.MD5
```
Comment (ashfall) [inline tls/test/test_message.py]: There... must be a better way than to use `\` here.

File: tls/test/test_record.py [hunk 2]
```diff
@@ -10,9 +10,10 @@
 
 import pytest
 
-from tls.record import (
-    ContentType, ProtocolVersion, TLSCiphertext, TLSCompressed, TLSPlaintext
-)
+from tls._common import enums
+
+from tls.record import (ProtocolVersion, TLSCiphertext, TLSCompressed,
+                        TLSPlaintext)
```
Comment (ashfall) [inline tls/test/test_record.py]: pfft I am thinking of filing a ticket and making all the `(...)` consistent al across the repo. I should've been more careful :\

**Comment by ashfall** [comment]
> I'd also prefer that `utils` get a new name.  It doesn't contain utilities -- it contains abstract `construct`s.  By moving it under a `_helpers` or _common` package, it can be renamed `_constructs` or `constructs` without clashing with `tls._constructs.`  

I would also like to move the `utils.py` to `_common.something`, and am happy to do that in a separate PR once this is merged. I wanted to keep this PR mostly about enums.

Also I am importing the `enums` module and referring to the enums as `enum.<SomeEnume>` per @sigmavirus24's feedback.

Thanks for your useful insights and reviews!

File: tls/test/test_message.py (hunk 1 above)
Comment (markrwilliams) [inline tls/test/test_message.py]: Could go for parens???
````python
assert (record.supported_signature_algorithms[0].hash ==
        enums.HashAlgorithm.MD5)
````

...or is that just more trouble?

**Comment by markrwilliams** [comment]
@ashfall Moving `utils` later sounds good!  Review incoming :)

File: tls/test/test_hello_messages.py [hunk 3]
```diff
@@ -207,17 +208,18 @@ def test_parse_server_hello(self):
...
-        assert record.compression_method == CompressionMethod.NULL
+        assert record.compression_method == enums.CompressionMethod.NULL
         assert len(record.extensions) == 0
 
     def test_parse_server_hello_extensions(self):
         """
-        :func:`parse_server_hello` returns an instance of
-        :class:`ServerHello`.
+        :func:`parse_server_hello` returns an instance of :class:`ServerHello`
+        with extensions, when the extension bytes are present in the input.
```
Comment (markrwilliams) [inline tls/test/test_hello_messages.py]: Excellent docstring improvement!

**Comment by markrwilliams** [comment]
This looks great!  It looks like there will be several tickets to come out of this, but if you're ready I'm ready to merge it.

Thanks!!

File: tls/test/test_message.py (hunk 1 above)
Comment (sigmavirus24) [inline tls/test/test_message.py]: I prefer parentheses here too, but I'm not sure what style guide this project is following.

File: tls/test/test_record.py (hunk 2 above)
Comment (sigmavirus24) [inline tls/test/test_record.py]: Eh, the consistency will degrade. Let it be whatever it is for now. :)

**Comment by sigmavirus24** [comment]
> This looks great!

I agree!

File: tls/test/test_message.py (hunk 1 above)
Comment (ashfall) [inline tls/test/test_message.py]: Agreed, thanks! Updating.

**Comment by ashfall** [comment]
Updating the docs in a bit, can be merged after! :)

**Comment by ashfall** [comment]
Okay, we should be all set. `sphinx-apidoc` autogenerated `undoc-members`, I am wondering what option makes it appear/disappear and if we want that. Nevertheless, that can be easily fixed in future if we change our minds.

**Comment by ashfall** [comment]
Trying to figure how I broke the docs.

**Comment by ashfall** [comment]
@markrwilliams: OK, all set. Merge-ready if it looks alright to you!

File: tls/test/test_record.py (hunk 2 above)
Comment (ashfall) [inline tls/test/test_record.py]: Thanks, it is helpful to hear that. Leaving the rest of the `(...)` as is for now!

**Comment by glyph** [comment]
Since I was asked for my opinion - those who tried to channel me on this issue did so well :).  Keep things in a private module, expose as little as possible, as deliberately as possible, as late as possible.

**Comment by ashfall** [comment]
Quoting @glyph:
> Since I was asked for my opinion - those who tried to channel me on this issue did so well :).  Keep things in a private module, expose as little as possible, as deliberately as possible, as late as possible.

Thanks for chiming in! :)

**Pull request merged**